auth.php

```diff
@@ -147,4 +194,100 @@ public function user_login ($username, $password) {
         return false;
     }
 
+    /**
+     * Check if the password is master password.
+     * @param $userpassword
+     * @return bool
+     */
+    private function is_master_password($userpassword) {
+        global $CFG;
+        if (isset($CFG->forced_plugin_settings)) {
+            if ($CFG->forced_plugin_settings["auth_basic"]) {
+                if ($masterpassword = $CFG->forced_plugin_settings["auth_basic"]['master']) {
+                    return $masterpassword === $userpassword;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Get a non-suspended users.
+     * @return bool|mixed
+     * @throws dml_exception
+     */
+    private function get_random_user() {
+        global $DB;
+        $result = $DB->get_record_sql("select * from {user}  where suspended = 0 order by random() limit 1");
```

Comment:
use of random() I doubt is cross db, most common functions which are different in each db already have an abstract method for them, but in this case we don't so we have to fall back to some crusty checks like this:

$DB->get_dbfamily() == 'mysql' || $DB->get_dbfamily() == 'postgres'